File: 해시테이블/캐시/캐시.py
from collections import OrderedDict

def solution(cacheSize, cities):
    time, cache = 0, OrderedDict()
    
    for city in map(lambda x: x.lower(), cities):
        if city in cache:
            cache.pop(city)
            cache[city] = True
            time += 1
        else:
            cache[city] = True
            if len(cache) > cacheSize: cache.popitem(last=False)
            time += 5
    return time

File: 해시테이블/캐시/캐시.java
import java.util.LinkedList;
import java.util.HashSet;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        LinkedList<String> recent = new LinkedList();
        HashSet<String> db = new HashSet();
        int answer = 0;
        
        for(String city : cities) {
            city = city.toUpperCase();
            
            if(db.contains(city)) {
                answer += 1;
                recent.remove(city);
                recent.offerFirst(city);
            }
            else {
             	answer += 5;
                db.add(city);
                recent.offerFirst(city);
                if(db.size() > cacheSize)
                    db.remove(recent.removeLast());
            }
        }
        return answer;
    }
}
